File: taxiwaydriver/lib/core/state/auth_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../api/api_client.dart';
import '../data/auth_repository.dart';
import '../models/driver_profile.dart';

class AuthState {
  final String phone;
  final DriverProfile? driver;
  final String? debugOtp;

  /// True right after a Google sign-in whose account has no phone on file
  /// yet. While true, the phone/OTP screens attach a phone number to the
  /// already-authenticated Google session instead of starting a fresh login.
  final bool linking;

  const AuthState({this.phone = '', this.driver, this.debugOtp, this.linking = false});

  bool get isLoggedIn => driver != null;
}

class AuthController extends Notifier<AuthState> {
  @override
  AuthState build() => const AuthState();

  Future<String?> sendOtp(String phone) async {
    state = AuthState(phone: phone, driver: state.driver, linking: state.linking);
    final debugOtp = state.linking
        ? await ref.read(authRepositoryProvider).sendPhoneLinkOtp(phone)
        : await ref.read(authRepositoryProvider).sendOtp(phone);
    state = AuthState(phone: phone, driver: state.driver, debugOtp: debugOtp, linking: state.linking);
    return debugOtp;
  }

  Future<void> verifyOtp(String code) async {
    if (state.linking) {
      final driver = await ref.read(authRepositoryProvider).confirmPhoneLink(state.phone, code);
      state = AuthState(phone: state.phone, driver: driver);
      return;
    }

    final driver = await ref.read(authRepositoryProvider).verifyOtp(state.phone, code);
    state = AuthState(phone: state.phone, driver: driver);
  }

  /// Signs in with a Firebase ID token obtained via [GoogleAuthService].
  /// Returns whether the account already has a phone linked — false means
  /// the caller should collect one through the normal phone/OTP screens,
  /// which [sendOtp]/[verifyOtp] now handle in "linking" mode.
  Future<bool> loginWithGoogle(String idToken) async {
    final result = await ref.read(authRepositoryProvider).loginWithGoogle(idToken);
    state = AuthState(
      phone: result.driver.phone,
      driver: result.phoneLinked ? result.driver : null,
      linking: !result.phoneLinked,
    );
    return result.phoneLinked;
  }

  void logout() {
    ApiClient.instance.clearToken();
    state = const AuthState();
  }
}

final authControllerProvider = NotifierProvider<AuthController, AuthState>(AuthController.new);

File: taxiway_admin/app/Http/Controllers/Api/Customer/AuthController.php
<?php

namespace App\Http\Controllers\Api\Customer;

use App\Http\Controllers\Controller;
use App\Models\Customer;
use App\Services\FirebaseTokenVerifier;
use App\Services\InvalidFirebaseTokenException;
use App\Services\OtpService;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;

class AuthController extends Controller
{
    public function __construct(
        private readonly OtpService $otp,
        private readonly FirebaseTokenVerifier $firebase,
    ) {}

    public function sendOtp(Request $request): JsonResponse
    {
        $data = $request->validate([
            'phone' => ['required', 'digits:10'],
        ]);

        $otp = $this->otp->generateAndSend('customer', $data['phone']);

        return response()->json([
            'message' => 'OTP sent.',
            ...($otp['debug'] ? ['debug_otp' => $otp['code']] : []),
        ]);
    }

    public function verifyOtp(Request $request): JsonResponse
    {
        $data = $request->validate([
            'phone' => ['required', 'digits:10'],
            'code' => ['required', 'digits:6'],
        ]);

        if (! $this->otp->verify('customer', $data['phone'], $data['code'])) {
            return response()->json(['message' => 'Invalid or expired OTP.'], 422);
        }

        $isNewCustomer = ! Customer::where('phone', $data['phone'])->exists();

        $customer = Customer::firstOrCreate(
            ['phone' => $data['phone']],
            ['name' => 'New Customer']
        );

        $token = $customer->createToken('device')->plainTextToken;

        return response()->json([
            'data' => [
                'customer' => $customer,
                'is_new_customer' => $isNewCustomer,
                'token' => $token,
            ],
        ]);
    }

    /**
     * Google Sign-In (Firebase Auth) login. Looks the customer up by the
     * Firebase uid, falling back to a phone-flow account with a matching
     * email so the two identities merge instead of duplicating. Phone stays
     * the account's canonical identity — a fresh Google account is created
     * with no phone yet, and `phone_linked` tells the app whether it still
     * needs to route the user through the OTP screen once to attach one.
     */
    public function google(Request $request): JsonResponse
    {
        $data = $request->validate([
            'id_token' => ['required', 'string'],
        ]);

        try {
            $profile = $this->firebase->verify($data['id_token']);
        } catch (InvalidFirebaseTokenException) {
            return response()->json(['message' => 'Invalid Google sign-in token.'], 422);
        }

        $customer = Customer::where('google_id', $profile['uid'])->first();

        if (! $customer && $profile['email']) {
            $customer = Customer::where('email', $profile['email'])->first();
        }

        if ($customer) {
            $customer->fill(array_filter([
                'google_id' => $profile['uid'],
                'photo_url' => $customer->photo_url ?: $profile['picture'],
            ]))->save();
        } else {
            $customer = Customer::create([
                'name' => $profile['name'] ?? 'New Customer',
                'email' => $profile['email'],
                'google_id' => $profile['uid'],
                'photo_url' => $profile['picture'],
            ]);
        }

        $token = $customer->createToken('device')->plainTextToken;

        return response()->json([
            'data' => [
                'customer' => $customer,
                'token' => $token,
                'phone_linked' => $customer->phone !== null,
            ],
        ]);
    }

    /** Sends an OTP to attach a phone number to the authenticated customer's account. */
    public function sendPhoneLinkOtp(Request $request): JsonResponse
    {
        $data = $request->validate([
            'phone' => ['required', 'digits:10'],
        ]);

        $otp = $this->otp->generateAndSend('customer', $data['phone']);

        return response()->json([
            'message' => 'OTP sent.',
            ...($otp['debug'] ? ['debug_otp' => $otp['code']] : []),
        ]);
    }

    /** Verifies the OTP and attaches the phone number to the authenticated customer. */
    public function verifyPhoneLink(Request $request): JsonResponse
    {
        $data = $request->validate([
            'phone' => ['required', 'digits:10'],
            'code' => ['required', 'digits:6'],
        ]);

        if (! $this->otp->verify('customer', $data['phone'], $data['code'])) {
            return response()->json(['message' => 'Invalid or expired OTP.'], 422);
        }

        $customer = $request->user();

        $takenByAnotherAccount = Customer::where('phone', $data['phone'])
            ->where('id', '!=', $customer->id)
            ->exists();

        if ($takenByAnotherAccount) {
            return response()->json(['message' => 'This phone number is already linked to another account.'], 422);
        }

        $customer->update(['phone' => $data['phone']]);

        return response()->json(['data' => ['customer' => $customer]]);
    }

    /** Registers/refreshes the FCM token for the authenticated customer's current device. */
    public function updateDeviceToken(Request $request): JsonResponse
    {
        $data = $request->validate([
            'fcm_token' => ['required', 'string'],
        ]);

        $request->user()->update(['fcm_token' => $data['fcm_token']]);

        return response()->json(['message' => 'Device token registered.']);
    }
}
